gh-108337: Add pyatomic.h header

colesbury opened the pull request:
This adds a new header that provides atomic operations on common data types.

Implementing PEP 703 requires use of atomic operations on more data types than provided by pycore_atomic.h. Additionally, pycore_atomic.h is only usable from Py_BUILD_CORE modules; it can't be used in public headers. PEP 703 will require atomic operations in object.h for Py_INCREF/DECREF, for example. The intention is that this will be exposed through Python.h, although that is not the case yet.

To avoid build issues in third-party extensions, the `pyatomic.h` header generally does not require `-std=gnu11` or `-std=c11` to be passed to the compiler (for GCC or Clang). When compiling C, MSVC will use the `pyatomic_msc.h`, which uses compiler intrinsics. When compiled in C++ mode, MSVC will use the `pyatomic_std.h` implementation, which uses C++11 atomics.

<!-- gh-issue-number: gh-108337 -->
* Issue: gh-108337
<!-- /gh-issue-number -->

colesbury committed:
gh-108337: Add pyatomic.h header

This adds a new header that provides atomic operations on common data
types. The intention is that this will be exposed through Python.h,
although that is not the case yet. The only immediate use is in
the test file.

colesbury committed:
Fix blurb name

colesbury committed:
Move pyatomic_std/gcc/msc.h to Include/cpython

Placates Tools/c-analyzer/check-c-globals.py.

colesbury committed:
Fix _InterlockedCompareExchange64 calls on MSVC x86

colesbury committed:
Move pyatomic.h to Include/cpython

colesbury committed:
Remove volatile from function signature.

Instead use volatile casts in MSVC implementation where they are
meaningful.

colesbury committed:
Add code documentation to pyatomic.h

colesbury committed:
Fix typo

colesbury committed:
Revert inadvertent change to stable_abi.py

vstinner commented on Include/cpython/pyatomic.h:
```diff
@@ -0,0 +1,397 @@
+#ifndef Py_ATOMIC_H
+#define Py_ATOMIC_H
+
+// This header provides cross-platform low-level atomic operations
```
You can move this help at the top, before the #ifndef/#define dance.

vstinner commented on Include/cpython/pyatomic.h:
```diff
@@ -0,0 +1,397 @@
+#define _Py_USE_GCC_BUILTIN_ATOMICS 1
+#endif
+#endif
+#endif
```
Indent for readability:

```suggestion
#ifndef _Py_USE_GCC_BUILTIN_ATOMICS
#  if defined(__GNUC__) && (__GNUC__ > 4 || (__GNUC__ == 4 && __GNUC_MINOR__ >= 8))
#    define _Py_USE_GCC_BUILTIN_ATOMICS 1
#  elif defined(__clang__)
#    if __has_builtin(__atomic_load)
#      define _Py_USE_GCC_BUILTIN_ATOMICS 1
#    endif
#  endif
#endif
```

vstinner commented on Include/cpython/pyatomic.h:
```diff
@@ -0,0 +1,397 @@
+#include "cpython/pyatomic_msc.h"
+#else
+#error "define pyatomic for this platform"
+#endif
```
Indent + change error message:

```suggestion
#if _Py_USE_GCC_BUILTIN_ATOMICS
#  define Py_ATOMIC_GCC_H
#  include "cpython/pyatomic_gcc.h"
#elif __STDC_VERSION__ >= 201112L && !defined(__STDC_NO_ATOMICS__)
#  define Py_ATOMIC_STD_H
#  include "cpython/pyatomic_std.h"
#elif defined(_MSC_VER)
#  define Py_ATOMIC_MSC_H
#  include "cpython/pyatomic_msc.h"
#else
#  error "no available pyatomic implementation for this platform/compiler"
#endif
```

vstinner commented on Include/cpython/pyatomic_gcc.h:
```diff
@@ -0,0 +1,679 @@
+#  error "this header file must not be included directly"
+#endif
+
+// This is the implementation of Python atomic operations using GCC's built-in
```
I suggest moving it at the top of the file.

vstinner commented on Include/cpython/pyatomic_gcc.h:
```diff
@@ -0,0 +1,679 @@
+_Py_atomic_add_int(int *address, int value)
+{
+    return __atomic_fetch_add(address, value, __ATOMIC_SEQ_CST);
+}
```
Usually, I prefer verbose syntax like the one that you used. But in this header file, you have tons of static inline functions, so I suggest using the compact syntax:

```suggestion
static inline int
_Py_atomic_add_int(int *address, int value)
{ return __atomic_fetch_add(address, value, __ATOMIC_SEQ_CST); }
```

vstinner commented on Include/cpython/pyatomic.h:
```diff
@@ -0,0 +1,397 @@
+
+// Atomically adds `value` to `address` and returns the previous value
+static inline int
+_Py_atomic_add_int(int *address, int value);
```
The glib library uses *atomic* for the first argument. Maybe it's a better name than *address* or *pointer*, what do you think?

colesbury replied:
I slightly prefer *address* or *ptr* rather than *atomic*

vstinner replied:
* https://en.cppreference.com/w/c/atomic uses ``obj`` name.
* https://www.ibm.com/docs/en/zos/2.1.0?topic=c11-atomic-load also uses *obj* name (C11 functions)
* pycore_atomic.h uses ``ATOMIC_VAL`` name
* pycore_atomic_func.h uses ``var`` name

I dislike the ``address`` name. What you pass is not a ``void*`` pointer or an ``uintptr_t`` address, but a reference to an atomic variable. I suggest to use *atomic* or *obj* name.

For add operation, the C11 API uses *arg* for the second parameter name. I'm fine with value.

colesbury replied:
I don't think `obj` makes as much sense in this API because, unlike the C11 API, they are not pointers to atomic objects (e.g., `_Atomic type`). They are just pointers to plain C variables. I think that's an important distinction.

If you dislike `address`, I can change it to `ptr`, which we commonly use and is a bit more concise.

colesbury replied:
For context, the GCC documentation for built-in atomic functions uses `ptr`. This is the API that `pyatomic.h` most closely matches.

https://gcc.gnu.org/onlinedocs/gcc/_005f_005fatomic-Builtins.html

vstinner commented on Include/cpython/pyatomic_gcc.h:
```diff
@@ -0,0 +1,679 @@
+static inline int
+_Py_atomic_compare_exchange_int(int *address, int expected, int value)
+{
+    return __atomic_compare_exchange_n(address, &expected, value, 0, __ATOMIC_SEQ_CST, __ATOMIC_SEQ_CST);
```
Can you try to git into 80 columns? (Write this call in two lines.) Same remark for code below.

vstinner commented on Include/cpython/pyatomic_gcc.h:
```diff
@@ -0,0 +1,679 @@
+_Py_atomic_compare_exchange_ptr(void *address, void *expected, void *value)
+{
+    void *e = expected;
+    return __atomic_compare_exchange_n((void **)address, &e, value, 0, __ATOMIC_SEQ_CST, __ATOMIC_SEQ_CST);
```
Wait, that's strange. address type should be ``void**``, no? Can you explain this code?

pitrou commented on Include/cpython/pyatomic.h:
```diff
@@ -0,0 +1,397 @@
+
+// Performs an atomic compare-and-exchange. If `*address` and `expected` are equal,
+// then `value` is stored in `*address`. Returns 1 on success and 0 on failure.
+// These correspond to the "strong" variations of the C11 atomic_compare_exchange_* functions.
```
C11 passes `expected` as a pointer, so that it's updated with the actual value when the latter doesn't match the former. Why not keep that convention here?

colesbury replied:
The motivation was two-fold: First, many of the _Py_atomic_compare_exchange calls in the nogil fork use constants as `expected` and this would be more verbose if it needs to be a pointer, e.g.:

```
_Py_atomic_compare_exchange_uint8(&m->v, LOCKED, UNLOCKED)
```

vs.

```
uint8_t expected = LOCKED:
_Py_atomic_compare_exchange_uint8(&m->v, &expected, UNLOCKED)
```

Second, I find this style (no pointer for expected) to be a bit less error-prone. I've been tripped up once or twice by having `expected` be modified when I didn't expect it.

I don't feel terribly strongly about this, so if there is a general preference for sticking closer to the C11-style API here, I can change it.

pitrou replied:
AFAIU the main motivation for the C11 style APIs is for retry loops in lockless data structure implementations. A simplistic example (this may be embarassingly wrong):
```c

struct ListNode;

typedef struct ListNode {
  int value;
  struct ListNode* next;
} ListNode;

void ListAppend(ListNode* list, int new_value) {
  ListNode* new_node = (ListNode*) malloc(sizeof ListNode);
  new_node->value = new_value;
  new_node->next = NULL;
  ListNode* expected = NULL;
  while (_Py_atomic_compare_exchange_ptr(&list->next, &expected, new_node)) {
    list = expected;
    expected = NULL;
  }
}
```

vstinner replied:
If the second argument is a reference (``&expected``), does it mean that it changes the value of the second argument (``*expected``)?

For C11 atomic_exchange(), the second argument is not a pointer, but a value (integer), no? https://en.cppreference.com/w/c/atomic/atomic_exchange 

C11 atomic_compare_exchange_strong() and atomic_compare_exchange_weak() use a pointer for expected. But this API writes into ``*expected`` if the ``*obj`` is not equal to ``*expected``.

> The behavior of atomic_compare_exchange_* family is as if the following was executed atomically:

```c
if (memcmp(obj, expected, sizeof *obj) == 0) {
    memcpy(obj, &desired, sizeof *obj);
    return true;
} else {
    memcpy(expected, obj, sizeof *obj);
    return false;
}
```

For this header fie, I would prefer to not have two flavors, the API is already quite long! I would prefer to have a single flavor. If there is an usecase where setting expected is relevant, I suggest to use a pointer for the second argument.

In short, I agree to change the API to ``int _Py_atomic_compare_exchange_int32(int32_t *obj, int32_t *expected, int32_t desired)``. The behavior should be well documented.

*obj*, *expected* and *desired* names come from the C11 API.

colesbury replied:
I'll make the second argument a pointer like the C11 API.

pitrou commented on Include/cpython/pyatomic.h:
```diff
@@ -0,0 +1,397 @@
+_Py_atomic_add_uintptr(uintptr_t *address, uintptr_t value);
+
+static inline Py_ssize_t
+_Py_atomic_add_ssize(Py_ssize_t *address, Py_ssize_t value);
```
Not sure it's worth exposing atomic ops for all int sizes and signednesses?

colesbury replied:
I expect to use at least one atomic operation on each of the data types here (but not every atomic op on every data type). I tried to be consistent on what's defined because it makes understanding what's available easier and testing easier.

pitrou commented on Include/cpython/pyatomic.h:
```diff
@@ -0,0 +1,397 @@
+_Py_atomic_compare_exchange_ssize(Py_ssize_t *address, Py_ssize_t expected, Py_ssize_t value);
+
+static inline int
+_Py_atomic_compare_exchange_ptr(void *address, void *expected, void *value);
```
Should it be `void** address` for clarity?

colesbury replied:
The problem is that `void **` requires an explicit cast for almost every use, because things like `PyObject **` are not implicitly convertible to `void **`. In other words, currently we can write things like:

```
PyObject *old_exc = _Py_atomic_exchange_ptr(&tstate->async_exc, exc);
```

but if `address` was `void **address`, we'd have to write:

```
PyObject *old_exc = _Py_atomic_exchange_ptr((void **)&tstate->async_exc, exc);
```

vstinner replied:
A large part of the Python C API uses macro to convert arguments to ``PyObject*``. Would it make sense to do the same here?

```
#define _Py_atomic_exchange_ptr(atomic, value) _Py_atomic_exchange_ptr(_Py_CAST(void**, atomic), (value))
```

colesbury replied:
I don't see the benefit of that style over the current approach, and it would silently allow passing some integer types to `_Py_atomic_exchange_ptr`.

vstinner replied:
Right, in that case I'm fine with the surprising ``void*`` type.

vstinner commented on Include/cpython/pyatomic.h:
```diff
@@ -0,0 +1,397 @@
+
+// Release fence
+static inline void
+_Py_atomic_fence_release(void);
```
```suggestion
// Sequential consistency fence
static inline void _Py_atomic_fence_seq_cst(void);

// Release fence
static inline void _Py_atomic_fence_release(void);
```

pitrou commented on Include/cpython/pyatomic.h:
```diff
@@ -0,0 +1,397 @@
+_Py_atomic_store_uint64_release(uint64_t *address, uint64_t value);
+
+static inline void
+_Py_atomic_store_ptr_release(void *address, void *value);
```
I'm not an expert, but why is it useful to expose "release" operations if no "acquire" operations are exposed?

colesbury replied:
You can always use stronger orderings for correctness (i.e., "seq_cst" everywhere instead of "acquire"). From a performance view, "release" is substantially faster than "seq_cst" stores on x86/x86-64, but "acquire" generates the same code as "seq_cst" loads on both x86/x86-64 and aarch64.

pitrou replied:
I don't think CPython support is limited to x86 and ARM variants, so the set of atomic ops exposed should probably be made consistent nevertheless?

Also, using "seq_cst" in combination with "release" will probably make the code more difficult to reason about, than if "acquire" is exposed (and memory ordering is already hard to reason about!).

colesbury replied:
I'll add a `_Py_atomic_load_ptr_acquire` for consistency (and I think I can remove the `_Py_atomic_store_uint64_release`).

vstinner commented on Include/cpython/pyatomic_msc.h:
```diff
@@ -0,0 +1,860 @@
+
+static inline int
+_Py_atomic_add_int(int *address, int value)
+{
```
Maybe add ``Py_BUILD_ASSERT(sizeof(int) == sizeof(long));``, since the cast is surprising.

Same remark for other casts below:

```
sizeof(uint8_t) == sizeof(char)
sizeof(uint16_t) == sizeof(short)
...
```

vstinner commented on Include/cpython/pyatomic_msc.h:
```diff
@@ -0,0 +1,860 @@
+#if defined(_M_X64) || defined(_M_ARM64)
+    return (int64_t)_InterlockedExchangeAdd64((volatile __int64*)address, (__int64)value);
+#else
+    for (;;) {
```
pycore_atomic.h uses a different style which may be more readable:

``do { ... } while (CMPXHG(...) == old_value);``

colesbury replied:
This can be written as:

1. (current version) or equivalently with `while (1)`
```
    for (;;) {
        __int64 old_value = *(volatile __int64*)address;
        __int64 new_value = old_value + (__int64)value;
        if (old_value == _InterlockedCompareExchange64((volatile __int64*)address, new_value, old_value)) {
            return old_value;
        }
    }
```

2. (equivalent but with do {} while)
```
    do {
        __int64 old_value = *(volatile __int64*)address;
        __int64 new_value = old_value + (__int64)value;
        if (old_value == _InterlockedCompareExchange64((volatile __int64*)address, new_value, old_value)) {
            return old_value;
        }
    } while(1);
```

3. `do { ... } while (CMPXHG(...) == old_value);`
```
    __int64 old_value;
    do {
        old_value = *(volatile __int64*)address;
        __int64 new_value = old_value + (__int64)value;
    } while (old_value != _InterlockedCompareExchange64((volatile __int64*)address, new_value, old_value));
    return old_value;
```

I find (3) a bit harder to read and I like when the variables modified in the loop are scoped to the loop (as in 1 and 2), but if there is a strong preference for this style I will change it.

vstinner replied:
I'm fine with the current style (1) ``for (;;)``.

vstinner commented on Include/cpython/pyatomic_msc.h:
```diff
@@ -0,0 +1,860 @@
+static inline uint64_t
+_Py_atomic_add_uint64(uint64_t *address, uint64_t value)
+{
+    return (uint64_t)_Py_atomic_add_int64((volatile int64_t*)address, (int64_t)value);
```
There is a compiler warning: ``different 'volatile' qualifiers``. _Py_atomic_add_int64() doesn't need volatile.

colesbury replied:
Thanks, I'll fix this.

vstinner commented on Include/cpython/pyatomic_msc.h:
```diff
@@ -0,0 +1,860 @@
+static inline uint64_t
+_Py_atomic_exchange_uint64(uint64_t *address, uint64_t value)
+{
+    return (uint64_t)_Py_atomic_exchange_int64((volatile __int64*)address, (__int64)value);
```
same compiler warning ``different 'volatile' qualifiers``, remove volatile? also cast to int64_t, not __int64.

vstinner commented on Include/cpython/pyatomic_msc.h:
```diff
@@ -0,0 +1,860 @@
+static inline void
+_Py_atomic_store_uint64(uint64_t *address, uint64_t value)
+{
+    _Py_atomic_exchange_int64((volatile __int64*)address, (__int64)value);
```
same compiler warning: remove volatile?

vstinner commented on Include/cpython/pyatomic.h:
```diff
@@ -0,0 +1,397 @@
+_Py_atomic_store_ptr_release(void *address, void *value);
+
+
+// Sequential consistency fence
```
I would prefer a more elaborated documentation, "fence" is kind of weak.

[__faststorefence doc](https://learn.microsoft.com/en-us/cpp/intrinsics/faststorefence?view=msvc-170):

> Guarantees that every previous memory reference, including both load and store memory references, is globally visible before any subsequent memory reference.

[_ReadWriteBarrier doc](https://learn.microsoft.com/en-us/cpp/intrinsics/readwritebarrier?view=msvc-170):

> Limits the compiler optimizations that can reorder memory accesses across the point of the call.

[dmb ish](https://devblogs.microsoft.com/oldnewthing/20220812-00/?p=106968):

> The data memory barrier ensures that all preceding writes are issued before any subsequent memory operations (including speculative memory access).

colesbury replied:
The challenge I have is that it's really hard to describe what fences do in a way that's helpful and accurate. The above documentation is too strong for C11 fences.

There's https://en.cppreference.com/w/c/atomic/atomic_thread_fence, but I find it vague. And the C++ documentation (https://en.cppreference.com/w/cpp/atomic/atomic_thread_fence) is more detailed but really hard to understand.

vstinner replied:
Please add one or two of these links here. It's ok to have references to external doc, it's better than no doc :-)

vstinner reviewed (changes requested):
It's unclear to me if volatile must be used on the first parameter in the function definition, or not. For me, it's surprising to be able to cast ``int*`` to ``volatile long*`` in ``pyatomic_msc.h``. Maybe explain the black magic in the documentation at the top of  pyatomic.h?

I'm scared by the _ptr variant which takes a ``void*`` and is then casted to a pointer of a pointer (``void**``). It looks suspicious.

colesbury committed:
Update Include/cpython/pyatomic.h

Co-authored-by: Victor Stinner <vstinner@python.org>

colesbury committed:
Update Include/cpython/pyatomic.h

Co-authored-by: Victor Stinner <vstinner@python.org>

colesbury committed:
Update Include/cpython/pyatomic.h

Co-authored-by: Victor Stinner <vstinner@python.org>

colesbury committed:
Move code documentation around.

Fix volatile warning in MSVC.

colesbury committed:
Clean-up pyatomic_msc.h and add Py_BUILD_ASSERT

colesbury committed:
Add _Py_atomic_load_ptr_acquire.

And remove _Py_atomic_store_uint64_release.

colesbury committed:
Format files

colesbury committed:
Add note about atomics in whatsnew

colesbury committed:
Describe volatile meaning on MSVC.

vstinner reviewed (commented):
The glib library provides g_atomic_int_dec_and_test() to implement reference counting. How would you reimplement it with your API? I'm not used to atomic variables and I never know how to use them *correctly*.

colesbury committed:
Add link to atomic_thread_fence documentation

colesbury committed:
Fix _Py_atomic_exchange_int64 on 32-bit Windows

vstinner commented on PCbuild/pythoncore.vcxproj.filters:
```diff
@@ -423,6 +423,12 @@
+    <ClInclude Include="..\Include\cpython\pyatomic.h">
+      <Filter>Include</Filter>
+    </ClInclude>
+    <ClInclude Include="..\Include\cpython\pyatomic_msc.h">
+      <Filter>Include</Filter>
+    </ClInclude>
```
You should add GCC here:

```suggestion
    <ClInclude Include="..\Include\cpython\pyatomic_gcc.h">
      <Filter>Include</Filter>
    </ClInclude>
    <ClInclude Include="..\Include\cpython\pyatomic_msc.h">
      <Filter>Include</Filter>
    </ClInclude>
```

It's just for the UI, not to build Python.

vstinner reviewed (commented):
We should decide _Py_atomic_compare_exchange() second argument should be a pointer or not. Apparently, a pointer covers more cases and so should be used.

colesbury committed:
Add pyatomic_gcc/std.h to vcxproj filters

colesbury committed:
Changes from review:

- Add pyatomic_*.h headers to MSVC filters
- Use a pointer for 2nd argument of compare_exchange functions
- Rename address to ptr

vstinner commented on Include/cpython/pyatomic.h:
```diff
@@ -0,0 +1,406 @@
+// equal, then `desired` is stored in `*ptr`. Otherwise `*expected` is updated
+// with the current value of `*ptr`. Returns 1 on success and 0 on failure.
+// These correspond to the "strong" variations of the C11
+// atomic_compare_exchange_* functions.
```
There is a single function, no?

```suggestion
// These correspond to the C11
// atomic_compare_exchange_strong() function.
```

atomic_compare_exchange_weak_explicit() has 2 additional arguments (of type memory_order).

colesbury committed:
Update Include/cpython/pyatomic.h

Co-authored-by: Victor Stinner <vstinner@python.org>

colesbury committed:
Rename ptr to obj
